### DISCUSSION TITLE
How do we upload images inside Rich Text Editor (AKA: Document Field) ?

### DISCUSSION THREAD

**Discussion by siamekanto19** [keystonejs/keystone#7237] [Questions] [answered]
Uploading images inside Rich Text Editors is very common in most other Headless CMS. But I can't seem to find the option to add Image Uploading functionality inside the Rich Text Editor of Keystonejs. Is there a way that I am not aware of or is it really not possible here?

**Comment by CanRau** [comment]
Also curious how that could work in Keystone 6, in v5 there was a Cloudinary package which also had a component block so it could be integrated with the former document field. Though I have no idea how to make a [Component Block](https://keystonejs.com/docs/guides/document-fields#component-blocks) so that it can have side effects to upload files to some place (I would go with Backblaze's S3 compatible API)?
Uuh looks like it would need a component block with a [custom field](https://keystonejs.com/docs/guides/document-fields#fields) I guess 🤔

Any hint would be highly appreciated 🙏

**Reply by CanRau** [reply]
I kinda got something working uploading to Backblaze though I had to hard code environment variables as I've no idea how to read them, and also would be much better if I could figure out how to use the Keystone server to generate a pre-signed upload URL or even proxy the download as to not expose API keys¿?

Okay found a way to [extend-express-app](https://keystonejs.com/docs/apis/config#extend-express-app), so now working on an endpoint for the upload which can also take care of the env vars

Edit: got it working 😻 next up polishing and blog post

**Comment by musps** [comment]
Any solution?

**Reply by CanRau** [reply]
It's still very much a work in progress though adding images inline works already and could be tweaked to allow any kind of file
[How to add images to WYSIWYG in Keystone v6](https://www.canrau.com/en/how-to-add-images-to-wysiwyg-in-keystone-v6) I'm going keep the article updated as I advance. Can also be adapted to save files to disk instead of external services. I'm thinking about were to best optimize images to have different sizes, could be on upload or later, probably doing it on upload for now.
Happy about feedback

**Reply by gmjacob** [reply]
Found this very helpful, thanks!

**Reply by hhimanshu** [reply]
@CanRau, do you have a public repo that I can look at?

**Reply by hhimanshu** [reply]
I am trying to reproduce what you did at https://github.com/hhimanshu/keystone-custom-document-fields/pull/1

Your help to open and make it work is appreciated and will also help others in the Keystone community. 
Thanks @CanRau

**Reply by hhimanshu** [reply]
@CanRau , what is running on your `127.0.0.1:8787`? I see two references in your [blog post](https://www.canrau.com/en/how-to-add-images-to-wysiwyg-in-keystone-v6) about this

```
<img width="200" src={`http://127.0.0.1:8787/${file}`} key={file} />
```

**Reply by CanRau** [reply]
Sorry for the late response, it's running a Cloudflare Worker locally using wrangler which returns the corresponding file from a private Backblaze B2 bucket. So depending on your needs and setup you provide the URL to your file backend, could be minio, AWS S3 or whatever 😉
Hope that helps, right now I don't have time to look into your PR sorry, but I try to respond here a little quicker if you've got more questions.

Maybe could update my article to make that more clear when I find the time

**Reply by hhimanshu** [reply]
No worries, I will ask here.
So, when someone uploads a file, `/api/upload` is called and uploads the file to Backblaze server. Then, at the same time, you request the same file from server using `127.0.0.1:8787` (this is your proxy to Backblaze via Cloudflare). Is that understanding correct?

**Reply by CanRau** [reply]
Not exactly "at the same time" though as only files are being requested which you uploaded before that's why the proxy URL contains the `${file}` at the end

**Comment by raveling** [comment, marked as answer]
Just to confirm the workarounds discussed here are currently the best way to get images in the `document` field. 

We are working on a more seamless story for Keystone images in general that will have a positive impact on this topic as well.

**Reply by hhimanshu** [reply]
@raveling , as you are thinking in this direction, please include images, videos (as uploads, and embeds - YouTube, Vimeo) in your design considerations.
As you guys build on Slate, it would be nice to see how you can leverage what they provide already

https://www.slatejs.org/examples/images
https://www.slatejs.org/examples/embeds
Thanks

**Reply by hhimanshu** [reply]
Also, can I subscribe to the story the team is working on? Are there any planned milestones to achieve that?

**Reply by raveling** [reply]
> please include images, videos (as uploads, and embeds - YouTube, Vimeo) in your design considerations.

these are already under consideration ✅ 

> can I subscribe to the story the team is working on? Are there any planned milestones to achieve that?

currently our [roadmap](https://keystonejs.com/updates/roadmap) and announcements in the community Slack are the best resources for updates.

hope this helps!

**Reply by limegorilla** [reply]
@raveling Are there any updates to this? Currently there isn't anything in the roadmap that I can see that relate to this.

**Comment by proteye** [comment]
Hi all!
Here is my example of image uploader in document editor (Document Field): https://gist.github.com/proteye/b9fc465caeea474fef3873b7297d58ad

**Comment by alihesari** [comment]
I discovered an excellent example that demonstrates how to implement custom fields and user interfaces for uploading images, YouTube videos, Tweets, and more. You can find it here: [KeystoneJS Document Field Customisation Example](https://github.com/keystonejs/keystone/tree/main/examples/document-field-customisation). This resource is particularly useful for those looking to extend the functionality of KeystoneJS in their projects.
